janus-idp/operator
# Add error messages to backstage CR status

## Description
When the operator fails to reconcile a CR, it should present an error in the CR's status for troubleshooting. 
When a CR has been successfully, the CR status should indicate that the deployment process is complete.

Simplified CR status conditions to use Deployed only.

Note that the running status support needs more investigation (on the possible impact on performance) and will be postponed to a future CR.

## Which issue(s) does this PR fix or relate to

- Fixes #71 (partially)

## PR acceptance criteria

- [ X] Tests

## How to test changes / Special notes to the reviewer
Deploy example/bs1.yaml and the follow CR status should be displayed indicating sync completion.
```
status:
  conditions:
  - lastTransitionTime: "2024-01-23T15:06:21Z"
    message: ""
    reason: DeployOK
    status: "True"
    type: Deployed
``` 

Now update the CR, set application replicas to -1 and then apply the CR change:
```
apiVersion: janus-idp.io/v1alpha1
kind: Backstage
metadata:
  name: bs1
spec:
  application:
    replicas: -1
``` 
You should see the CR status like below:
```
status:
  conditions:
  - lastTransitionTime: "2024-01-23T15:08:24Z"
    message: 'failed to deploy Backstage Deployment: Deployment.apps "backstage-bs1"
      is invalid: spec.replicas: Invalid value: -1: must be greater than or equal
      to 0'
    reason: DeployFailed
    status: "False"
    type: Deployed
``` 

Now if you revert the CR to the default:
```
apiVersion: janus-idp.io/v1alpha1
kind: Backstage
metadata:
  name: bs1
spec:
``` 
You should see DeployOK again:
```
status:
  conditions:
  - lastTransitionTime: "2024-01-23T15:09:06Z"
    message: ""
    reason: DeployOK
    status: "True"
    type: Deployed
```
